docs/proposals/blocks-storage-sharding.md

```diff
@@ -0,0 +1,84 @@
+
+In Cortex, when using the experimental blocks storage, each querier internally runs the Thanos [`BucketStore`](https://github.com/thanos-io/thanos/blob/master/pkg/store/bucket.go). This means that each querier has a full view over all blocks in the long-term storage and all blocks index headers are loaded in each querier memory. The querier memory usage linearly increase with number and size of all blocks in the storage, imposing a scalability limit to the blocks storage.
+
+In this proposal we want to solve this. In particular, we want to:
+
+1. Shard blocks (index headers) across a pool of nodes
+2. Do not compromise HA on the read path (if a node fails, queries should continue to work)
+3. Do not compromise correctness (either the query result is correct or it fails)
+
+## Proposed solution
+
+The idea is to introduce a new Cortex service - `store-gateway` - internally running the Thanos [`BucketStore`](https://github.com/thanos-io/thanos/blob/master/pkg/store/bucket.go). At query time, a querier will run a query fetching the matching series both from ingesters and the subset of gateways holding the related blocks (based on the query time range). Blocks are replicated across the gateways in order to guarantee query results consistency and HA even in the case of a gateway instance failure.
+
+### Ring-based sharding and replication
+
+In order to build blocks sharding and replication, the `store-gateway` instances form a [ring](../architecture.md#the-hash-ring). Each gateway instance uses a custom [`MetaFetcherFilter`](https://github.com/thanos-io/thanos/blob/master/pkg/block/fetcher.go#L108) to filter blocks loaded on the instance itself, keeping only blocks whose `hash(block-id)` is within the tokens range assigned to the gateway instance within the ring.
+
+Within a gateway, the blocks synchronization is triggered in two cases:
+
+1. **Periodically**<br />
+   to discover new blocks uploaded by ingesters or compactor, and delete old blocks removed due to retention or by the compactor
+2. **On-demand**<br/>
+   when the ring topology changes (the tokens ranges assigned to the gateway instance have changed)
+
+It's important to outline that the sync takes time (typically will have to re-scan the bucket and download new blocks index headers) and Cortex needs to guarantee query results consistency at any given time (_see below_).
+
+### Query execution
+
+When a querier executes a query, it will need to fetch series both from ingesters and the store-gateway instances.
+
+For a given query, the number of blocks to query is expected to be low, especially if the Cortex cluster is running the `query-frontend` with a `24h` query split interval. In this scenario, whatever is the client's query time range, the `query-frontend` will split the client's query into sub-queries each with up to `24h` time range and the querier will likely hit not more than 1 block per sub-query (except for the last 24h for which blocks may have not been compacted yet).
```

Comment:
Let's not mix Prometheus `subqueries` with this (: I think we used other terms like `separate-queries` or something

Reply:
@owen-d What's the right terminology for the sub-interval queries done by the query-frontend?

Reply:
No one knows, but subqueries are wrong one. `Partitioned Query?`

Reply:
I think we call them `split` internally, but I'm not opinionated. `partitioned` sound good.

Reply:
My 2cents is go with `partitioned` here.

Reply:
Updated the doc using "partitioned query". Thanks for the feedback.